### PR TITLE
gnrc_ipv6: fix asserts for NHC

### PR DESCRIPTION
**Pull request by authmillenon**
Fixes problem described in https://github.com/RIOT-OS/RIOT/pull/5033#issuecomment-205610286. However with #5033 merged I still don't receive NHC compressed UDP packets (though the receiver isn't crashing anymore)

### CODE REVIEW

**Comment by Yonezawa-T2** [comment]
In general, `gnrc_ipv6_demux` may receive parsed packets like this:

```
GNRC_NETTYPE_UNDEF ← pkt
↓ next
GNRC_NETTYPE_YET_ANOTHER_AWESOME_PROTOCOL
↓ next
GNRC_NETTYPE_SOME_MARVELOUS_PROTOCOL_OVER_COAP
↓ next
GNRC_NETTYPE_COAP
↓ next
GNRC_NETTYPE_UDP ← current
↓ next
GNRC_NETTYPE_IPV6_EXT
↓ next
GNRC_NETTYPE_IPV6_EXT
↓ next
GNRC_NETTYPE_IPV6_EXT
↓ next
GNRC_NETTYPE_IPV6
↓ next
GNRC_NETTYPE_NETIF
```

So we should remove those assertions completely, and decide how to broadcast those packets.

Broadcasting `current` with `GNRC_NETTYPE_UDP` would be wrong, since receiver cannot access UDP payload. Broadcasting `pkt` with `GNRC_NETTYPE_UDP` may be fine for UDP, but if `current` is `GNRC_NETTYPE_IPV6_EXT`, broadcasting `pkt` with `GNRC_NETTYPE_IPV6_EXT` is ambiguous.

A possible strategy is the following, but I'm not sure it is correct:

- Broadcast `current` with `current->type` if `current->type` is `GNRC_NETTYPE_IPV6_EXT`
- Broadcast `pkt` with `current->type` (not `pkt->type`) otherwise

Another strategy may be broadcasting `pkt` with `GNRC_NETTYPE_IPV6_EXT` only once after all extension headers are parsed, then receiver process all extension headers.

**Comment by authmillenon** [comment]
> In general, `gnrc_ipv6_demux` may receive parsed packets like this:
> 
> ```
> GNRC_NETTYPE_UNDEF ← pkt
> ↓ next
> GNRC_NETTYPE_YET_ANOTHER_AWESOME_PROTOCOL
> ↓ next
> GNRC_NETTYPE_SOME_MARVELOUS_PROTOCOL_OVER_COAP
> ↓ next
> GNRC_NETTYPE_COAP
> ↓ next
> GNRC_NETTYPE_UDP ← current
> ↓ next
> GNRC_NETTYPE_IPV6_EXT
> ↓ next
> GNRC_NETTYPE_IPV6_EXT
> ↓ next
> GNRC_NETTYPE_IPV6_EXT
> ↓ next
> GNRC_NETTYPE_IPV6
> ↓ next
> GNRC_NETTYPE_NETIF
> ```

In general yes, but in practice this shouldn't be a case that should happen, IMHO. For 6LoWPAN IPHC it is a special case, because we don't want to copy around stuff in that adaption layer for decoding, but why would an already parsed application layer protocol packet land here again unless something went horrible wrong. Side note: application layer stuff shouldn't be implemented GNRC internally.

> So we should remove those assertions completely, and decide how to broadcast those packets.
>
> Broadcasting current with GNRC_NETTYPE_UDP would be wrong, since receiver cannot access UDP payload. Broadcasting pkt with GNRC_NETTYPE_UDP may be fine for UDP, but if current is GNRC_NETTYPE_IPV6_EXT, broadcasting pkt with GNRC_NETTYPE_IPV6_EXT is ambiguous.
> 
> A possible strategy is the following, but I'm not sure it is correct:
> 
> Broadcast current with current->type if current->type is GNRC_NETTYPE_IPV6_EXT
Broadcast pkt with current->type (not pkt->type) otherwise
Another strategy may be broadcasting pkt with GNRC_NETTYPE_IPV6_EXT only once after all extension headers are parsed, then receiver process all extension headers.

**Comment by Yonezawa-T2** [comment]
> What do you mean with "broadcasting"?

I mean `_dispatch_rcv_pkt`.

> in practice this shouldn't be a case that should happen

I see. Then, do we need to change only following lines to handle UDP as a special case?
```
        _dispatch_rcv_pkt(current->type, GNRC_NETREG_DEMUX_CTX_ALL, current);
        _dispatch_rcv_pkt(GNRC_NETTYPE_IPV6, nh, current);
```

**Comment by authmillenon** [comment]
Done.

**Comment by Yonezawa-T2** [comment]
We still have extension headers:

```
GNRC_NETTYPE_UDP ← pkt
↓ next
GNRC_NETTYPE_IPV6_EXT
↓ next
GNRC_NETTYPE_IPV6_EXT ← current
↓ next
GNRC_NETTYPE_IPV6_EXT
↓ next
GNRC_NETTYPE_IPV6
↓ next
GNRC_NETTYPE_NETIF
```

Dispatching `pkt` with `GNRC_NETTYPE_IPV6_EXT` is ambiguous. Dispatching `pkt` with `pkt->type` is wrong.

One strategy is dispatching `current` with `current->type` like the old implementation.
Another strategy is dispatching `pkt` with `GNRC_NETTYPE_IPV6_EXT` only once after all extension headers are parsed, then receiver process all extension headers.

**Comment by authmillenon** [comment]
Okay, can you maybe provide a patch for that to this PR? It seems your fixes made IPv6 not as clear cut as it used to be... :-/

**Comment by Yonezawa-T2** [comment]
Made a patch. Please cherry-pick.

https://github.com/Yonezawa-T2/RIOT/tree/gnrc_ipv6_handle_nhc_packet
https://github.com/Yonezawa-T2/RIOT/commit/9d3320b412b92ed77c593843091f1b3d792fff68

# Dispatched Events

- `GNRC_NETTYPE_IPV6_EXT with GNRC_NETREG_DEMUX_CTX_ALL` is dispatched only once.
- `GNRC_NETTYPE_IPV6` is dispatched for each next headers.

Example:

Suppose we have following packet:
```
GNRC_NETTYPE_UNDEF ← pkt
↓ next
GNRC_NETTYPE_UDP
↓ next
GNRC_NETTYPE_IPV6_EXT (nh = PROTNUM_IPV6_EXT_RH)
↓ next
GNRC_NETTYPE_IPV6_EXT (nh = PROTNUM_IPV6_EXT_HOPOPT) ← current
↓ next
GNRC_NETTYPE_IPV6
↓ next
GNRC_NETTYPE_NETIF
```

Following events will be dispatched:

- `GNRC_NETTYPE_IPV6_EXT` with `GNRC_NETREG_DEMUX_CTX_ALL`
- `GNRC_NETTYPE_IPV6` with `PROTNUM_IPV6_EXT_HOPOPT`
- `GNRC_NETTYPE_IPV6` with `PROTNUM_IPV6_EXT_RH`
- `GNRC_NETTYPE_UDP` with `GNRC_NETREG_DEMUX_CTX_ALL`
- `GNRC_NETTYPE_IPV6` with `PROTNUM_UDP`


# netapi.h and gnrc_netapi.c

Added `release` flag to `gnrc_netapi_dispatch`. If `release` is `true`, it releases the packet after dispatching the packet. Otherwise, it keeps the packet. So we can eliminate ugly `_dispatch_rcv_pkt`.

# gnrc_ipv6.c

Removed `_dispatch_rcv_pkt` in favor of `gnrc_netapi_dispatch` with `release` flag `false`.

Fixed an old bug:
```diff
-    pkt->type = gnrc_nettype_from_protnum(nh);
+    current->type = gnrc_nettype_from_protnum(nh);
```

# gnrc_ipv6_ext.c

Scans all extension header in `gnrc_ipv6_ext_demux` and dispatches `GNRC_NETTYPE_IPV6` with each extension headers.

# Test

`tests/gnrc_ipv6_ext` now checks `gnrc_ipv6.c` with unparsed UDP packet and parsed UDP packet. It works as expected.

**Comment by authmillenon** [comment]
Can you maybe open a PR to my branch? This would make it much easier (and comprehensible) to comment on your changes.

**Comment by Yonezawa-T2** [comment]
Created https://github.com/authmillenon/RIOT/pull/28

**Comment by authmillenon** [comment]
Thanks.

**Comment by authmillenon** [comment]
Reverted the changes to netapi and used a `gnrc_pktbuf_hold()` prior to the call to dispatch instead. Please check if everything is still working properly.

**Comment by authmillenon** [comment]
Seems like this PR is necessary for NHC to work now, so I re-tagged it for the next release.

**Comment by Yonezawa-T2** [comment]
Using `gnrc_pktbuf_hold` instead of adding `release` flag is much better idea.

File: sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c [hunk 1]
```diff
@@ -100,86 +150,66 @@ void gnrc_ipv6_ext_demux(kernel_pid_t iface,
...
+                /* TODO: add handling of types */
+                nh = ext->nh;
+                DEBUG("ipv6_ext: next header = %" PRIu8 "\n", nh);
+
+                if ((current = _mark_extension_header(current, &pkt)) == NULL) {
+                    return;
+                }
+
+                gnrc_pktbuf_hold(pkt, 1);   /* don't release on next dispatch */
+                gnrc_netapi_dispatch(GNRC_NETTYPE_IPV6, nh, GNRC_NETAPI_MSG_TYPE_RCV, pkt, false);
```
Comment (Yonezawa-T2) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: `gnrc_netapi_dispatch(GNRC_NETTYPE_IPV6, nh, GNRC_NETAPI_MSG_TYPE_RCV, pkt);`

Comment (Yonezawa-T2) [inline sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c]: and if `gnrc_netapi_dispatch` returns false, `gnrc_pktbuf_release(pkt)`.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 2]
```diff
@@ -100,6 +97,54 @@ kernel_pid_t gnrc_ipv6_init(void)
...
+                              (!interested);
+
+        if (!should_release) {
+            gnrc_pktbuf_hold(pkt, 1);   /* don't remove from packet buffer in
+                                         * next dispatch */
+        }
+        if (!gnrc_netapi_dispatch(current->type,
+                                  GNRC_NETREG_DEMUX_CTX_ALL,
+                                  GNRC_NETAPI_MSG_TYPE_RCV,
+                                  pkt)) {
```
Comment (Yonezawa-T2) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]:
```diff
diff --git a/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c b/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c
index fe838e6..173f97c 100644
--- a/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c
+++ b/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c
@@ -124,6 +124,11 @@ static bool _dispatch_next_header(gnrc_pktsnip_t *current,
                                   GNRC_NETREG_DEMUX_CTX_ALL,
                                   GNRC_NETAPI_MSG_TYPE_RCV,
                                   pkt)) {
+            gnrc_pktbuf_release(pkt);
+
+            if (should_release) {
+                return false;
+            }
         }
 
         if (should_release) {
@@ -138,8 +143,12 @@ static bool _dispatch_next_header(gnrc_pktsnip_t *current,
     if (!(gnrc_netapi_dispatch(GNRC_NETTYPE_IPV6,
                                nh,
                                GNRC_NETAPI_MSG_TYPE_RCV,
-                               pkt)) && (!interested)) {
-        return false;
+                               pkt))) {
+        gnrc_pktbuf_release(pkt);
+
+        if (!interested) {
+            return false;
+        }
     }
 
     return true;
```

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 3]
```diff
@@ -100,6 +97,54 @@ kernel_pid_t gnrc_ipv6_init(void)
     return gnrc_ipv6_pid;
 }
 
+static bool _dispatch_next_header(gnrc_pktsnip_t *current,
```
Comment (Yonezawa-T2) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Sorry, returning `bool` was redundant. We might save a few bytes by changing it to `void`.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 4]
```diff
@@ -100,6 +97,54 @@ kernel_pid_t gnrc_ipv6_init(void)
...
+                                  gnrc_pktsnip_t *pkt,
+                                  uint8_t nh,
+                                  bool interested)
+{
+#ifdef MODULE_GNRC_IPV6_EXT
+    const bool current_is_ipv6 = ((current->type != GNRC_NETTYPE_IPV6_EXT) ||
+                                  (current->next->type == GNRC_NETTYPE_IPV6));
+#else
+    const bool current_is_ipv6 = (current->next->type == GNRC_NETTYPE_IPV6);
```
Comment (Yonezawa-T2) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: `current_is_ipv6` doesn't reflect what it is. `should_dispatch_current_type` or something would be better.

**Comment by Yonezawa-T2** [comment]
I made a private branch reflecting all comments.
https://github.com/Yonezawa-T2/RIOT/tree/gnrc_ipv6_handle_nhc_packet_2
https://github.com/Yonezawa-T2/RIOT/commit/85c98d5b07e9cc7ffd2c6c2724a06e98784f6f7b
Confirmed `tests/gnrc_ipv6_ext` works. I will ACK when cherry-picked.

**Comment by authmillenon** [comment]
We've decided to revert NHC for now instead, but we will try to backport it into the release as soon as it is out.

**Comment by authmillenon** [comment]
> I made a private branch reflecting all comments.
https://github.com/Yonezawa-T2/RIOT/tree/gnrc_ipv6_handle_nhc_packet_2
Yonezawa-T2@85c98d5
Confirmed tests/gnrc_ipv6_ext works. I will ACK when cherry-picked.

Done and squashed.

**Comment by authmillenon** [comment]
@cgundogan please retest with your setup, too.

**Comment by Yonezawa-T2** [comment]
ACK and go when CI is Happy.

**Comment by authmillenon** [comment]
Let's please wait, if multihop-NHC is working too again :-)

**Comment by authmillenon** [comment]
(i.e. second ACK needed by @cgundogan)

**Comment by authmillenon** [comment]
@cgundogan assured me offline that he will have tested it until the start of next week.

**Comment by PeterKietzmann** [comment]
And did @cgundogan manage to do so :-) ?

**Comment by cgundogan** [comment]
tested UDP with activated NHC over multiple-hops. Looks good - ACK. Merge once Murdock agrees.

**Comment by authmillenon** [comment]
Backport provided in #5402

**Pull request merged**